File: Clone Graph - Leetcode 133/Clone Graph - Leetcode 133.cpp
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <stack>
using namespace std;


class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (!node) return nullptr;

        unordered_map<Node*, Node*> oToN;
        stack<Node*> stk;
        unordered_set<Node*> visited;

        stk.push(node);
        visited.insert(node);

        // DFS to create the nodes
        while (!stk.empty()) {
            Node* curr = stk.top();
            stk.pop();
            oToN[curr] = new Node(curr->val);

            for (Node* nei : curr->neighbors) {
                if (visited.find(nei) == visited.end()) {
                    visited.insert(nei);
                    stk.push(nei);
                }
            }
        }

        // DFS to set up neighbors
        for (auto& entry : oToN) {
            Node* oldNode = entry.first;
            Node* newNode = entry.second;
            for (Node* nei : oldNode->neighbors) {
                newNode->neighbors.push_back(oToN[nei]);
            }
        }

        return oToN[node];
    }
};

File: Clone Graph - Leetcode 133/Clone Graph - Leetcode 133.js
/**
 * Definition for a Node.
 * function Node(val, neighbors) {
 *     this.val = val;
 *     this.neighbors = neighbors === undefined ? [] : neighbors;
 * }
 */

/**
 * @param {Node} node
 * @return {Node}
 */
var cloneGraph = function(node) {
    if (!node) return null;

    const oToN = new Map();
    const stk = [node];
    const visited = new Set();

    visited.add(node);

    // DFS to create the nodes
    while (stk.length > 0) {
        const curr = stk.pop();
        oToN.set(curr, new Node(curr.val));

        for (const nei of curr.neighbors) {
            if (!visited.has(nei)) {
                visited.add(nei);
                stk.push(nei);
            }
        }
    }

    // DFS to set up neighbors
    for (const [oldNode, newNode] of oToN.entries()) {
        for (const nei of oldNode.neighbors) {
            newNode.neighbors.push(oToN.get(nei));
        }
    }

    return oToN.get(node);
};
